### ISSUE TITLE
Fink 0.45.5: "Problem with MergeList"/"Problem parsing dependency Pre-Depends" after update

### ISSUE THREAD

**Issue by lucajovine** [fink/fink#240]
Hello, after updating to 0.45.5 on a machine that is still running Mac OS 10.13.6 (due to some legacy software not compatible with 10.14+), I am getting the following issues with update/dist-upgrade/upgrade/check:

> sudo apt-get -fy update
(...)
Hit http://bindist.finkproject.org stable/main Release
Reading Package Lists...E: Problem parsing dependency Pre-Depends
E: Error occured while processing cffi-py27 (NewVersion1)

E: Problem with MergeList /sw/var/lib/apt/lists/_sw_fink_dists_stable_main_binary-darwin-x86%5f64_Packages
E: The package lists or status file could not be parsed or opened.

> sudo apt-get -fy dist-upgrade
Reading Package Lists...E: Problem parsing dependency Pre-Depends
E: Error occured while processing cffi-py27 (NewVersion1)
E: Problem with MergeList /sw/var/lib/apt/lists/_sw_fink_dists_stable_main_binary-darwin-x86%5f64_Packages
E: The package lists or status file could not be parsed or opened.

etc.

The MergeList file mentioned above does exist and is both readable and populated. Moreover, if I restore my /sw directory from a backup and run the above commands as well as fink update-all without ever issuing fink-selfupdate, everything works just fine... so it appears to be an issue with the 0.45.5 update itself.

Any idea of how to fix this?

Thanks, Luca

**Comment by TheSin-** [comment]
yes it forsake can't be empty, so we need to add something in fink to stop that, but how did it get into the bindists?  Those haven't been updated for a long time.

**Comment by jrandall814** [comment]
I run fink 45.5 on 10.13.6 but don’t know how to test your fix.
On ‘fink selfupdate’ my install reports this parse error for the whatever package in /opt/sw/var/lib/apt/lists/_opt_sw_fink_dists_stable_main_binary-darwin-x86%5f64_Packages first contains a line with an empty Pre-Depends:  value. l edited that local file and deleted all of those lines, but the file is simply replaced as it was by the download from Hit http://bindis.finkmirrors.net stable/main Packages
during the next execution of ‘fink selfupdate’ (or ‘fink scanpackages’).  I searched my local fink code for a perl parse function upon which I could perform an edit recommended by a user of debian apt-get, but couldn’t find one.

**Comment by TheSin-** [comment]
that file isn't downloaded that is locally built, you need to mod a different file but it's not save hence why I'm not posting it.

Once the fix is added I'll write a one line script to fix your local file.

**Comment by TheSin-** [comment]
This will fix your install but be very very careful with it.

```
sudo cp /opt/sw/var/lib/status /opt/sw/var/lib/status.old
sudo sed -i -e '/^Pre-Depends: *$/d' /opt/sw/var/lib/status
fink index
```

Run this only once and if it brakes then run this to revert it

```
sudo cp /opt/sw/var/lib/status.old /opt/sw/var/lib/status
fink index
```

Again I warn, this is dangerous so be very careful when modifying the status file.

This will only fix it till you build a new package then it will break again.  The proposed fix the fink needs to be tested and a new version of fink needs to be installed for a perm fix.

**Comment by jrandall814** [comment]
I tried editing my /opt/sw/lib/perl5/Fink/PkgVersion.pm as you first suggested and just re-running ‘fink selfupdate’, but that was to no effect.
Then I tried the temporary fix you givevabove (my status file was in /opt/sw/var/lib/dpkg). I still got the same error on my first new run of ‘fink selfupdate’

**Comment by TheSin-** [comment]
that Fix won't repair the damage it'll just prevent it from happening

**Comment by TheSin-** [comment]
the key to not allow black Pre-Depends: lines at all in the status or packages files.  Currently it could occur, it's not happening to me cause all my package depend on dpkg (>= 1.15). but when using dpkg 1.10 it'll always be blank with the current code base.  SO my patch corrects that and only that.  Then fixing the status file which is the script I sent you.  I might have to run it on packages too, BUT I thought they would regain with the status file.

I'm not sure right now so do not do it, but the package files might be able to be removed and reindexed once the status file is clean.  I don't have a dpkg 1.10 install so I haven't been able to test so it's all just theory ATM.

**Comment by nieder** [comment]
@TheSin-: editing PkgVersion.pm (dpkg1.10) makes `.../binary-darwin-x86_64/Packages.gz` no longer add an empty `Pre-Depends:` field to packages. So thats seems good for fixing it moving forward.

But the status file doesn't have Pre-Depends listed for any package, regardless of the patch status of PkgVersion.pm.
I'm skeptical about editing `.../binary-darwin-x86_64/Packages.gz` since it's gzipped. If I move it out of the way, will it be rebuilt? Do I need to run `fink scanpackages` to do that?

**Comment by TheSin-** [comment]
I thought index would, sadly scabpackes will re read the debs which means you’ll need to rebuild all the debs that have bad control files :( you might need to make a list from the Packages.gj and issue a finkrebuild on that list now that you are patched

**Comment by TheSin-** [comment]
A new release of fink should go out right away!

**Comment by nieder** [comment]
```
$ zcat /sw/fink/dists/stable/main/binary-darwin-x86_64/Packages.gz | grep -B 10 Pre-Depends | grep "^Package: " | cut -f 2 -d: | wc -l
    1293
```

Oof.

**Comment by nieder** [comment]
Some more testing:
The files that are actually breaking things are `/sw/var/lib/apt/lists/_sw_fink_dists_stable_main_binary-darwin-x86%5f64_Packages`

I think this can be fixed by the hard way by:
1) Patching PkgVersion.pm
2) running `fink cleanup --all` to remove old debs and clean up the "status" database
3) Running this command (change `/sw` at the beginning to your Fink install path)
```
for i in `ls -1 /sw/var/lib/apt/lists/_*Packages`; do grep -B 10 Pre-Depends $i | grep "^Package: " | cut -f 2 -d: | xargs echo; done
```
4) running `fink rebuild <the long list of packages from step 3>`
5) run `fink scanpacakges; sudo apt-get update` so that the apt lists files are updated.

I'm working on way to possible get the deb filenames instead and remove those.

**Comment by TheSin-** [comment]
We really should accept my patch and push a new version of fink ASAP though to stop it from getting worse in the meantime.

**Comment by dmacks** [comment]
@TheSin's patch solves the immediate problem of blank Pre-Depends fields in .deb. Committed as f49f1db4.

There is an unrelated logical bug in this region of code, which will lead to a blank Depends: if the .info has a Pre-Depends: that contains the kernel-version. But I don't think .info Pre-Depends is fully supported by the dependency engine so I'm comfortable getting this hot-fix out ASAP.

**Comment by nieder** [comment]
I've released 0.45.6 that fixes the underlying cause.
To remove broken debs, I believe the following will do (change /opt/sw to whatever your Fink prefix might be)

```
cd /opt/sw
# this gives you a list of all the .debs in question in case you want to save any
for baddeb in `ls -1 var/lib/apt/lists/_*Packages`; do grep -A 4 Pre-Depends $baddeb | grep "^Filename: " | cut -f 2 -d: | sort; done
# this will delete the list above. It will ask for your password to delete the files. BE CAREFUL
for baddeb in `ls -1 var/lib/apt/lists/_*Packages`; do grep -A 4 Pre-Depends $baddeb | grep "^Filename: " | cut -f 2 -d: | sort | sudo rm -f $baddeb done
fink scanpackages
sudo apt-get update
```

This will not affect your installed packages since the problem was with the .deb archives and not the installed package itself.

**Comment by lucajovine** [comment]
> I've released 0.45.6 that fixes the underlying cause. To remove broken debs, I believe the following will do (change /opt/sw to whatever your Fink prefix might be)
> 
> ```
> cd /opt/sw
> # this gives you a list of all the .debs in question in case you want to save any
> for baddeb in `ls -1 var/lib/apt/lists/_*Packages`; do grep -A 4 Pre-Depends $baddeb | grep "^Filename: " | cut -f 2 -d: | sort; done
> # this will delete the list above. It will ask for your password to delete the files. BE CAREFUL
> for baddeb in `ls -1 var/lib/apt/lists/_*Packages`; do grep -A 4 Pre-Depends $baddeb | grep "^Filename: " | cut -f 2 -d: | sort | sudo rm -f $baddeb; done
> fink scanpackages
> sudo apt-get update
> ```
> 
> This will not affect your installed packages since the problem was with the .deb archives and not the installed package itself.

Sorry for the silence during the last few days and thank you - I confirm that 0.45.6 fixed the issue on my system!

**Comment by TheSin-** [comment]
That will work, though it shouldn't be run if you have good debs with Pre-Depends like for those of us on the dpkg 1.16 branch, but for the master branch it'll work fine.
